### PR TITLE
[Plan Generation] Fetch group processing

### PR DESCRIPTION
**Pull request by SimonSapin**
Fixes https://github.com/apollographql/federation-next/issues/119 (mostly)

### CODE REVIEW

File: src/query_graph/graph_path.rs [hunk 1]
```diff
@@ -183,3 +184,22 @@ pub(crate) struct ClosedBranch(Vec<Arc<ClosedPath>>);
...
+            .map(|path_element| {
+                Ok(match path_element.as_ref() {
+                    OpPathElement::Field(field) => QueryPathElement::Field(field.try_into()?),
+                    OpPathElement::InlineFragment(inline) => {
+                        QueryPathElement::InlineFragment(inline.try_into()?)
+                    }
+                })
+            })
+            .collect()
+    }
```
Comment (SimonSapin) [inline src/query_graph/graph_path.rs:204]: Something cool but maybe not obvious: [`collect()`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.collect) can turn an iterator of results into a single result (stopping at the first error)

File: src/query_plan/fetch_dependency_graph.rs [hunk 2]
```diff
@@ -130,17 +135,68 @@ pub(crate) struct DeferTracking {
...
+        &self,
+        _config: &QueryPlannerConfig,
+        _handled_conditions: &Conditions,
+        _variable_definitions: &[Node<VariableDefinition>],
+        _fragments: Option<&RebasedFragments>,
+        _op_name: Option<String>,
+    ) -> Option<super::PlanNode> {
+        todo!()
+    }
+}
```
Comment (SimonSapin) [inline src/query_plan/fetch_dependency_graph.rs:172]: TODO

File: src/query_plan/fetch_dependency_graph_processor.rs [hunk 3]
```diff
@@ -1,2 +1,395 @@
...
+    config: Arc<QueryPlannerConfig>,
+    variable_definitions: Vec<Node<VariableDefinition>>,
+    fragments: Option<RebasedFragments>,
+    operation_name: Option<Name>,
+    assigned_defer_labels: Option<HashSet<NodeStr>>,
+    counter: u32,
+}
+
+// TODO
+pub(crate) struct RebasedFragments;
```
Comment (SimonSapin) [inline src/query_plan/fetch_dependency_graph_processor.rs:55]: TODO: what’s the right data structure for this?

File: src/query_plan/fetch_dependency_graph_processor.rs [hunk 4]
```diff
@@ -1,2 +1,395 @@
...
+        node: Option<PlanNode>,
+    ) -> Result<DeferredDeferBlock, FederationError> {
+        Ok(DeferredDeferBlock {
+            depends: defer_info
+                .dependencies
+                .iter()
+                .cloned()
+                .map(|id| DeferredDependency {
+                    id,
+                    defer_label: None, // TODO: is this never set?
```
Comment (SimonSapin) [inline src/query_plan/fetch_dependency_graph_processor.rs:290]: Unsure about this. In TypeScript it’s an implicit `undefined`.

File: src/query_plan/fetch_dependency_graph_processor.rs [hunk 5]
```diff
@@ -1,2 +1,395 @@
...
+            } else {
+                Some(defer_info.label.clone())
+            },
+            query_path: (&defer_info.path.full_path).try_into()?,
+            // Note that if the deferred block has nested @defer,
+            // then the `value` is going to be a `DeferNode`
+            // and we'll use it's own `subselection`, so we don't need it here.
+            sub_selection: if defer_info.deferred.is_empty() {
+                // TODO: port `withoutEmptyBranches`?
+                Some((&defer_info.sub_selection).try_into()?)
```
Comment (SimonSapin) [inline src/query_plan/fetch_dependency_graph_processor.rs:308]: TODO

File: src/query_plan/fetch_dependency_graph_processor.rs [hunk 6]
```diff
@@ -1,2 +1,395 @@
...
+    fn reduce_defer(
+        &mut self,
+        main: Option<PlanNode>,
+        sub_selection: &FetchSelectionSet,
+        deferred: Vec<DeferredDeferBlock>,
+    ) -> Result<Option<PlanNode>, FederationError> {
+        Ok(Some(PlanNode::Defer(Arc::new(DeferNode {
+            primary: PrimaryDeferBlock {
+                // TODO: port `withoutEmptyBranches`?
+                sub_selection: Some(sub_selection.selection_set.as_ref().try_into()?),
```
Comment (SimonSapin) [inline src/query_plan/fetch_dependency_graph_processor.rs:325]: TODO

File: src/query_plan/fetch_dependency_graph_processor.rs (hunk 4 above)
Comment (sachindshinde) [reply src/query_plan/fetch_dependency_graph_processor.rs:290]: @SimonSapin 
You're right, there's an entry for it in the TS for the node, but it's never set or read. Should be fine to remove it.

File: src/query_plan/fetch_dependency_graph.rs (hunk 2 above)
Comment (SimonSapin) [reply src/query_plan/fetch_dependency_graph.rs:174]: I’ve filed https://github.com/apollographql/federation-next/issues/158 to track following up on this.

File: src/query_plan/fetch_dependency_graph_processor.rs (hunk 5 above)
Comment (SimonSapin) [reply src/query_plan/fetch_dependency_graph_processor.rs]: Done

File: src/query_plan/fetch_dependency_graph_processor.rs (hunk 6 above)
Comment (SimonSapin) [reply src/query_plan/fetch_dependency_graph_processor.rs]: Done

File: src/query_plan/fetch_dependency_graph_processor.rs [hunk 7]
```diff
@@ -1,2 +1,398 @@
...
+/// sequential depending on whether we represent a `query` or a `mutation`),
+/// and the processor will be called on groups in such a way.
+pub(crate) trait FetchGroupProcessor<TProcessed, TDeferred> {
+    fn on_fetch_group(
+        &mut self,
+        group: &mut FetchDependencyGraphNode,
+        handled_conditions: &Conditions,
+    ) -> Result<TProcessed, FederationError>;
+    fn on_conditions(&mut self, conditions: &Conditions, value: TProcessed) -> TProcessed;
+    fn reduce_parallel(&mut self, values: &[TProcessed]) -> TProcessed;
```
Comment (sachindshinde) [inline src/query_plan/fetch_dependency_graph_processor.rs:97]: Something for later, but I think all the callsites of `reduce_parallel()` and `reduce_sequence()` will pass ownership of `TProcessed` (but cloning is cheap for `PlanNode`s, so this doesn't really matter).

**Review by sachindshinde — CHANGES_REQUESTED** [review]

File: src/query_plan/fetch_dependency_graph_processor.rs [hunk 8]
```diff
@@ -1,2 +1,398 @@
...
+    // work well in practice (and if it's a huge problem for someone, we can change it).
+    let mut chars = subgraph_name.chars().filter_map(|c| {
+        if let '-' | '_' = c {
+            Some('_')
+        } else {
+            c.is_ascii_alphanumeric().then_some(c)
+        }
+    });
+    let Some(first) = chars.next() else {
+        return "_".to_owned();
```
[the remark this answers is not in the document]
Comment (SimonSapin) [reply src/query_plan/fetch_dependency_graph_processor.rs:354]: Returning a string that is not a valid name from something called "to valid name" doesn’t feel right, so I’m changing it to return an option.

File: src/query_plan/fetch_dependency_graph_processor.rs (hunk 7 above)
Comment (SimonSapin) [reply src/query_plan/fetch_dependency_graph_processor.rs:97]: Changed to take iterators of owned `TProcessed` values.

File: src/query_plan/fetch_dependency_graph_processor.rs [hunk 9]
```diff
@@ -1,2 +1,398 @@
...
+}
+
+fn parallel_cost(values: &[QueryPlanCost]) -> QueryPlanCost {
+    values.iter().sum()
+}
+
+fn sequence_cost(values: &[QueryPlanCost]) -> QueryPlanCost {
+    // reduceRight((acc, stage, idx) => (acc + (Math.max(1, idx * pipeliningCost) * stage)), 0);
+    values
+        .iter()
```
[the remark this answers is not in the document]
Comment (sachindshinde) [reply src/query_plan/fetch_dependency_graph_processor.rs]: Good catch! You're right, `reduceRight` doesn't matter here since `+` is symmetric.

**Review by sachindshinde — APPROVED** [review]

**Pull request merged**